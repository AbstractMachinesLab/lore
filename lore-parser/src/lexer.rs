use logos::Logos;

#[derive(Logos, Debug, PartialEq)]
pub enum Token {
    #[regex("#.*(\r\n|\n)?", |lex| lex.slice()[1..].parse())]
    Comment(String),

    #[token("as")]
    As,

    #[token("prefix")]
    Prefix,

    #[token("using")]
    Using,

    #[token("kind")]
    Kind,

    #[token("attr")]
    Attribute,

    #[token("rel")]
    Relation,

    #[token("in")]
    In,

    #[token(":")]
    Colon,

    #[regex("(\"([^\"\\\\]|\\\\.)*\")", |lex| lex.slice()[1..lex.slice().len() -1].parse())]
    LiteralString(String),

    #[regex("(\"\"\"([^\"\\\\]|\\\\.)*\"\"\")", |lex| lex.slice()[3..lex.slice().len() -3].parse())]
    MultiLineLiteralString(String),

    #[token("{")]
    OpenBrace,

    #[token("}")]
    ClosedBrace,

    #[token("/")]
    Slash,

    #[regex("([a-z0-9][a-z0-9-]*:|@|:)[a-zA-Z0-9()+,-.:=@;$_!*'%/?#]+", |lex| lex.slice().parse())]
    URI(String),

    #[regex("[a-zA-Z][a-zA-Z0-9-_]*", |lex| lex.slice().parse())]
    Text(String),

    #[regex("[0-9]+", |lex| lex.slice().parse())]
    Number(u64),

    #[error]
    #[regex(r"[ \t\n\f]+", logos::skip)]
    Error,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn comment() {
        let mut lex = Token::lexer(r#"# spotify:artist:2Hkut4rAAyrQxRdof7FVJq "#);

        assert_eq!(
            lex.next(),
            Some(Token::Comment(
                " spotify:artist:2Hkut4rAAyrQxRdof7FVJq ".to_string()
            ))
        );
    }

    #[test]
    fn literal_string_single_line() {
        let mut lex = Token::lexer(r#" "spotify:artist:2Hkut4rAAyrQxRdof7FVJq" "#);

        assert_eq!(
            lex.next(),
            Some(Token::LiteralString(
                "spotify:artist:2Hkut4rAAyrQxRdof7FVJq".to_string()
            ))
        );
    }
    #[test]
    fn literal_string_multi_line() {
        let mut lex = Token::lexer(
            r#" """
            spotify:artist:2Hkut4rAAyrQxRdof7FVJq
            more lines
            """ "#,
        );

        assert_eq!(
            lex.next(),
            Some(Token::MultiLineLiteralString("\n            spotify:artist:2Hkut4rAAyrQxRdof7FVJq\n            more lines\n            ".to_string())));
    }

    #[test]
    fn uri() {
        let mut lex = Token::lexer(r#" spotify:artist:2Hkut4rAAyrQxRdof7FVJq "#);

        assert_eq!(
            lex.next(),
            Some(Token::URI(
                "spotify:artist:2Hkut4rAAyrQxRdof7FVJq".to_string()
            ))
        );
    }

    #[test]
    fn using_namespace() {
        let mut lex = Token::lexer(r#" using spotify:schema/2021 "#);

        assert_eq!(lex.next(), Some(Token::Using));
        assert_eq!(
            lex.next(),
            Some(Token::URI("spotify:schema/2021".to_string()))
        );
    }

    #[test]
    fn prefix_uri() {
        let mut lex = Token::lexer(r#" prefix spotify:schema/2021 as @spotify "#);

        assert_eq!(lex.next(), Some(Token::Prefix));
        assert_eq!(
            lex.next(),
            Some(Token::URI("spotify:schema/2021".to_string()))
        );
        assert_eq!(lex.next(), Some(Token::As));
        assert_eq!(lex.next(), Some(Token::URI("@spotify".to_string())));
    }

    #[test]
    fn prefixed_uri() {
        let mut lex = Token::lexer(r#" @spotify/schema/2021 "#);

        assert_eq!(
            lex.next(),
            Some(Token::URI("@spotify/schema/2021".to_string()))
        );
    }

    #[test]
    fn lowercase_kind_name() {
        let mut lex = Token::lexer(r#" kind u "#);

        assert_eq!(lex.next(), Some(Token::Kind));
        assert_eq!(lex.next(), Some(Token::Text("u".to_string())));
    }

    #[test]
    fn lowercase_attr_name() {
        let mut lex = Token::lexer(r#" attr u "#);

        assert_eq!(lex.next(), Some(Token::Attribute));
        assert_eq!(lex.next(), Some(Token::Text("u".to_string())));
    }

    #[test]
    fn empty_kind() {
        let mut lex = Token::lexer(r#" kind User "#);

        assert_eq!(lex.next(), Some(Token::Kind));
        assert_eq!(lex.next(), Some(Token::Text("User".to_string())));
    }

    #[test]
    fn empty_string() {
        let mut lex = Token::lexer(r#" "" "#);
        assert_eq!(lex.next(), Some(Token::LiteralString("".to_string())));
    }

    #[test]
    fn non_empty_string() {
        let mut lex = Token::lexer(r#" "what" "#);
        assert_eq!(lex.next(), Some(Token::LiteralString("what".to_string())));
    }

    #[test]
    fn kind_with_body() {
        let mut lex = Token::lexer(r#"kind User { }"#);

        assert_eq!(lex.next(), Some(Token::Kind));
        assert_eq!(lex.next(), Some(Token::Text("User".to_string())));
        assert_eq!(lex.next(), Some(Token::OpenBrace));
        assert_eq!(lex.next(), Some(Token::ClosedBrace));
    }

    #[test]
    fn empty_attribute() {
        let mut lex = Token::lexer(r#"attr Name in User"#);

        assert_eq!(lex.next(), Some(Token::Attribute));
        assert_eq!(lex.next(), Some(Token::Text("Name".to_string())));
        assert_eq!(lex.next(), Some(Token::In));
        assert_eq!(lex.next(), Some(Token::Text("User".to_string())));
    }

    #[test]
    fn field() {
        let mut lex = Token::lexer(r#"range: integer"#);

        assert_eq!(lex.next(), Some(Token::Text("range".to_string())));
        assert_eq!(lex.next(), Some(Token::Colon));
        assert_eq!(lex.next(), Some(Token::Text("integer".to_string())));
    }

    #[test]
    fn field_block() {
        let mut lex = Token::lexer(r#" { hello world } "#);

        assert_eq!(lex.next(), Some(Token::OpenBrace));
        assert_eq!(lex.next(), Some(Token::Text("hello".to_string())));
        assert_eq!(lex.next(), Some(Token::Text("world".to_string())));
        assert_eq!(lex.next(), Some(Token::ClosedBrace));
    }

    #[test]
    fn attr_with_fields() {
        let mut lex = Token::lexer(
            r#"
                attr Name {
                    @label/en "Name"
                    @label/es "Nombre"
                    @comment/en ""
                    @see-also    @other/entity

                    @symmetry       :symmetric
                    @reflexivity    :reflexive
                    @disjoint-with  :reflexive

                    @domain      User
                    @range       @lore/string
                    @cardinality 1
                }
        "#,
        );

        assert_eq!(lex.next(), Some(Token::Attribute));
        assert_eq!(lex.next(), Some(Token::Text("Name".to_string())));
        assert_eq!(lex.next(), Some(Token::OpenBrace));
        assert_eq!(lex.next(), Some(Token::URI("@label/en".to_string())));
        assert_eq!(lex.next(), Some(Token::LiteralString("Name".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@label/es".to_string())));
        assert_eq!(lex.next(), Some(Token::LiteralString("Nombre".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@comment/en".to_string())));
        assert_eq!(lex.next(), Some(Token::LiteralString("".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@see-also".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@other/entity".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@symmetry".to_string())));
        assert_eq!(lex.next(), Some(Token::URI(":symmetric".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@reflexivity".to_string())));
        assert_eq!(lex.next(), Some(Token::URI(":reflexive".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@disjoint-with".to_string())));
        assert_eq!(lex.next(), Some(Token::URI(":reflexive".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@domain".to_string())));
        assert_eq!(lex.next(), Some(Token::Text("User".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@range".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@lore/string".to_string())));
        assert_eq!(lex.next(), Some(Token::URI("@cardinality".to_string())));
        assert_eq!(lex.next(), Some(Token::Number(1)));
        assert_eq!(lex.next(), Some(Token::ClosedBrace));
    }
}
