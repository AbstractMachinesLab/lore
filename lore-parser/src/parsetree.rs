use lore_ast::URI;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Pos {
    file: String,
    line: u32,
    col: u32,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Name {
    URI(URI),
    Alias(String),
}

impl Into<lore_ast::Name> for &Name {
    fn into(self) -> lore_ast::Name {
        match self {
            Name::URI(uri) => lore_ast::Name::of_uri(&uri),
            Name::Alias(alias) => lore_ast::Name::unresolved_alias(alias),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum Literal {
    Number(u64),
    String(String),
    Name(Name),
}

#[derive(Clone, Debug, PartialEq)]
pub struct Field {
    pub name: Name,
    pub value: Literal,
}

#[derive(Clone, Debug, PartialEq)]
pub enum StructureItem {
    Namespace {
        uri: URI,
    },

    Comment(String),

    Alias {
        uri: URI,
        prefix: URI,
    },

    Kind {
        name: Name,
        fields: Vec<Field>,
    },

    Attribute {
        name: Name,
        fields: Vec<Field>,
    },

    Relation {
        subject: Name,
        predicate: Name,
        object: Name,
    },
}

#[derive(Clone, Debug, Default, PartialEq)]
pub struct Structure {
    items: Vec<StructureItem>,
}

impl Structure {
    pub fn of_items(items: Vec<StructureItem>) -> Structure {
        Structure { items }
    }

    pub fn items(&self) -> &Vec<StructureItem> {
        &self.items
    }
}
