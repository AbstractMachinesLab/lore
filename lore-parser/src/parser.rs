use crate::lexer::Token;
use crate::parsetree::*;
use logos::{Lexer, Logos};
use lore_ast::URI;
use thiserror::Error;

#[derive(Error, Debug, PartialEq, Eq)]
pub enum ParseError {
    #[error("Expected a URI.")]
    ExpectedURI,

    #[error("The `use` syntax expects a URI.")]
    UseExpectsURI,

    #[error(
        "The `use` syntax should follow the format: use <uri> as <name>. Did you forget the 'as'?"
    )]
    UseIsMissingTheAsKeyword,

    #[error(r#"
        The `use` syntax should follow the format: use <uri> as <name>. Did you forget to specify a name?
    "#)]
    UseIsMissingTheAliasedName,

    #[error("")]
    UseExpectsTextPrefix,

    #[error("The `kind <name>` syntax is missing a name.")]
    KindIsMissingAName,

    #[error("We expected to find a Name (an Alias or a URI) but found something else instead.")]
    NameIsInvalid,

    #[error("We expected to find a Name, did you forget it?")]
    NameIsMissing,

    #[error("The `rel` syntax should follow the format: rel <subject> <predicate> <object>. All 3 must be URIs or aliased names.")]
    RelationExpectedSubjectToBeName,

    #[error("The `rel` syntax should follow the format: rel <subject> <predicate> <object>. All 3 must be URIs or aliased names.")]
    RelationExpectedPredicateToBeName,

    #[error("The `rel` syntax should follow the format: rel <subject> <predicate> <object>. All 3 must be URIs or aliased names.")]
    RelationExpectedObjectToBeName,

    #[error("We expected to find an Alias, a Kind, an Attribute, or a Relation.")]
    ExpectedTopLevelItem,

    #[error("Runtime error")]
    Runtime(String),
}

pub fn parse(text: &str) -> Result<Structure, ParseError> {
    let mut lex = Token::lexer(text);
    parse_structure(&mut lex)
}

fn parse_structure(mut lex: &mut Lexer<Token>) -> Result<Structure, ParseError> {
    let mut items = vec![];
    while let Some(token) = lex.next() {
        let item = match token {
            Token::Use => parse_alias(&mut lex),
            Token::Kind => parse_kind(&mut lex),
            Token::Attribute => parse_attr(&mut lex),
            Token::Relation => parse_rel(&mut lex),
            _ => Err(ParseError::ExpectedTopLevelItem),
        }?;
        items.push(item);
    }
    Ok(Structure::of_items(items))
}

fn parse_uri(lex: &mut Lexer<Token>) -> Result<URI, ParseError> {
    match lex.next() {
        Some(Token::URI(uri)) => Ok(URI(uri)),
        _ => Err(ParseError::ExpectedURI),
    }
}

fn parse_name(lex: &mut Lexer<Token>) -> Result<Name, ParseError> {
    match lex.next() {
        Some(Token::URI(uri)) => Ok(Name::URI(URI(uri))),
        Some(Token::Text(text)) => Ok(Name::Alias(text)),
        None => Err(ParseError::NameIsMissing),
        _ => Err(ParseError::NameIsInvalid),
    }
}

fn parse_alias(mut lex: &mut Lexer<Token>) -> Result<StructureItem, ParseError> {
    let uri = match parse_uri(&mut lex) {
        Ok(uri) => Ok(uri),
        _ => Err(ParseError::UseExpectsURI),
    }?;

    match lex.next() {
        Some(Token::As) => Ok(()),
        _ => Err(ParseError::UseIsMissingTheAsKeyword),
    }?;

    let prefix = match lex.next() {
        Some(Token::Text(prefix)) => Ok(prefix),
        _ => Err(ParseError::UseIsMissingTheAliasedName),
    }?;

    Ok(StructureItem::Alias { uri, prefix })
}

fn parse_kind(mut lex: &mut Lexer<Token>) -> Result<StructureItem, ParseError> {
    let name = match parse_name(&mut lex) {
        Ok(name) => Ok(name),
        _ => Err(ParseError::KindIsMissingAName),
    }?;

    Ok(StructureItem::Kind { name })
}

fn parse_attr(mut lex: &mut Lexer<Token>) -> Result<StructureItem, ParseError> {
    let name = match parse_name(&mut lex) {
        Ok(name) => Ok(name),
        _ => Err(ParseError::KindIsMissingAName),
    }?;

    Ok(StructureItem::Attribute { name })
}

fn parse_rel(mut lex: &mut Lexer<Token>) -> Result<StructureItem, ParseError> {
    let subject = match parse_name(&mut lex) {
        Ok(name) => Ok(name),
        _ => Err(ParseError::RelationExpectedSubjectToBeName),
    }?;

    let predicate = match parse_name(&mut lex) {
        Ok(name) => Ok(name),
        _ => Err(ParseError::RelationExpectedPredicateToBeName),
    }?;

    let object = match parse_name(&mut lex) {
        Ok(name) => Ok(name),
        _ => Err(ParseError::RelationExpectedObjectToBeName),
    }?;

    Ok(StructureItem::Relation {
        subject,
        predicate,
        object,
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::*;

    macro_rules! test {
        ($name:ident, $src:expr) => {
            #[test]
            fn $name() {
                let result = parse($src);
                let snapshot = format!(
                    r#"
input:
    {}

output:

{:#?}
"#,
                    $src, result
                );
                assert_snapshot!(snapshot)
            }
        };
    }

    test!(
        parse_alias,
        " use spotify:artist:2Hkut4rAAyrQxRdof7FVJq as Rush "
    );

    test!(
        parse_alias_with_something_that_isnt_a_uri,
        " use 2Hkut4rAAyrQxRdof7FVJq as Rush "
    );

    test!(parse_alias_with_a_missing_uri, " use as Rush ");

    test!(
        parse_alias_with_uri_but_missing_keyword_as,
        " use spotify:hello Rush "
    );

    test!(
        parse_alias_with_uri_but_missing_the_aliased_name,
        " use spotify:hello as"
    );

    test!(parse_kind_with_uri_name, "kind spotify:kind:artist");

    test!(parse_kind_with_aliased_name, "kind Hello");

    test!(parse_kind_with_missing_name, "kind");

    test!(parse_attr_with_uri_name, "attr spotify:field:Name");

    test!(parse_attr_with_aliased_name, "attr Name");

    test!(parse_rel_incomplete_with_1_part_aliased, "rel Artist");

    test!(parse_rel_incomplete_with_1_part, "rel spotify:kinds/Artist");

    test!(parse_rel_incomplete_with_2_parts_aliased, "rel Artist has");

    test!(
        parse_rel_incomplete_with_2_parts,
        "rel spotify:kinds/Artist spotify:rels/has"
    );

    test!(
        parse_rel_complete_with_3_parts_aliased,
        "rel Artist has Name"
    );

    test!(
        parse_rel_complete_with_3_parts,
        "rel spotify:kinds/Artist spotify:rels/has spotify:attrs/Name"
    );

    test!(
        parse_multiple_items,
        r#"
            use spotify:kind:artist as Artist

            kind Artist

            attr Name

            attr spotify:field:play_count

            kind spotify:kind:Album

            use spotify:kind:song as Song

            kind Song
        "#
    );
}
