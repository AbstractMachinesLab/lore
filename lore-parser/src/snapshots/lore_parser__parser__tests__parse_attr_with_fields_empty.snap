---
source: lore-parser/src/parser.rs
expression: snapshot

---

input:
     attr Name {} 

output:

Ok(
    Structure {
        filename: "$name",
        items: [
            Attribute {
                name: Alias(
                    "Name",
                ),
                fields: [],
            },
        ],
    },
)

